R/deploymentTarget.R

```diff
@@ -1,93 +1,241 @@
+  envVars = NULL,
+  account = NULL,
+  server = NULL,
+  error_call = caller_env()
+) {
 
-    createDeploymentTarget(
-      appName = appName,
-      appTitle = appTitle,
-      appId = appId,
-      envVars = envVars,
-      username = fullAccount$name, # first deploy must be to own account
-      account = fullAccount$name,
-      server = fullAccount$server
+  # We must have a target account+server in order to use the appId.
+  # The selected account may not be the original creator of the content.
+  accountDetails <- accountInfo(account, server)
+
+  # Filtering is only by server and includes all deployments in case we have a deployment record
+  # from a collaborator.
+  appDeployments <- deployments(
+    appPath = recordPath,
+    serverFilter = server,
+    excludeOrphaned = FALSE
+  )
+  appDeployments <- appDeployments[appDeployments$appId == appId, ]
+  if (nrow(appDeployments) > 1) {
+    cli::cli_abort(
+      c(
+        "Supplied {.arg appId} ({appId}) identifies multiple deployments.",
+        i = "Manage obsolete deployments with rsconnect::forgetDeployment()."
+      ),
+      call = error_call
     )
-  } else if (nrow(appDeployments) == 1) {
-    # If both appName and appId supplied, check that they're consistent.
-    if (!is.null(appId) && appDeployments$appId != appId) {
-      cli::cli_abort(
-        c(
-          "Supplied {.arg appId} ({appId}) does not match deployment record ({appDeployments$appId}).",
-          i = "Omit {.arg appId} to use existing for deployment for app {.str {appName}}, or",
-          i = "Omit {.arg appName} to create new deployment record."
-        ),
-        call = error_call
-      )
-    }
+  }
 
-    updateDeploymentTarget(appDeployments, appTitle, envVars)
-  } else {
-    deployment <- disambiguateDeployments(appDeployments, error_call = error_call)
-    updateDeploymentTarget(deployment, appTitle, envVars)
+  # Existing local deployment record.
+  if (nrow(appDeployments) == 1) {
+    deployment <- appDeployments[1, ]
+    deployment <- updateDeploymentTarget(deployment, appTitle, envVars)
+    return(list(
+      accountDetails = accountDetails,
+      deployment = deployment
+    ))
   }
-}
 
-deploymentTargetForApp <- function(appId,
-                                   appTitle = NULL,
-                                   account = NULL,
-                                   server = NULL) {
-  accountDetails <- findAccount(account, server)
+  # No local deployment record. Get it from the server.
   application <- getApplication(accountDetails$name, accountDetails$server, appId)
 
-  createDeploymentTarget(
+  # Note: The account+server of this deployment record may
+  # not correspond to the original content creator.
+  deployment <- createDeploymentTarget(
     appName = application$name,
     appTitle = application$title %||% appTitle,
     appId = application$id,
-    envVars = NULL,
+    envVars = envVars,
     username = application$owner_username %||% accountDetails$name,
```

Comment:
Should we change the arguments to `createDeploymentTarget()` to more clearly differentiate the username of the current user vs. the username of the account owner? Might need to be in a separate PR to avoid cluttering this one, but I think it's likely to make understanding the code easier in the future.

Reply:
The deployment record has the account+server corresponding to local account record for the user responsible for creating the deployment record on disk. This is usually the creator, but not always (e.g. collaborator publish using `appId`). The username is associated with the creating user (unsure if we get this back from shinyapps.io).

I'm not sure how to best represent this possible inconsistency. Happy to file an issue for further discussion.

Reply:
I was thinking maybe we could rename `username` to something like `owner_account` to make it clear that this is also an account name, but it's the account name for the original author.